crates/noirc_evaluator/src/ssa/acir_gen/internal_var_cache.rs

```diff
@@ -76,10 +98,185 @@ impl InternalVarCache {
+        evaluator: &mut Evaluator,
+        ctx: &SsaContext,
+    ) -> Witness {
+        if let Some(ids) = ctx.constants.get(&value) {
+            //we have a constant node object for the value
+            if !ids.is_empty() {
+                let id = ids.first().unwrap();
+                let var = self.get_or_compute_internal_var_unwrap(*id, evaluator, ctx);
+                if let Some(w) = var.cached_witness() {
+                    return *w;
+                }
+            }
+            // We generate a witness and assigns it
+            let w = evaluator.create_intermediate_variable(Expression::from(value));
+            for &id in ids {
+                let mut cached_var = self.get_or_compute_internal_var_unwrap(id, evaluator, ctx);
+                assert!(cached_var.cached_witness().is_none());
+                cached_var.set_witness(Some(w));
+                self.update(cached_var);
+            }
+            w
+        } else {
+            //if not, we use the constants map
+            let var = self.constants.entry(value).or_insert(InternalVar::from_constant(value));
+            Self::const_to_witness_helper(var, evaluator)
+        }
+    }
+
+    // Helper function which generates a witness for an InternalVar
+    // Do not call outside const_to_witness()
+    fn const_to_witness_helper(var: &mut InternalVar, evaluator: &mut Evaluator) -> Witness {
+        let w = Self::internal_get_or_compute_witness(var, evaluator);
+        if w.is_none() {
+            let witness = expression_to_witness(var.to_expression(), evaluator);
+            var.set_witness(Some(witness));
+        }
+        var.cached_witness().unwrap()
+    }
+
+    /// Get or compute a witness for an internal var
+    /// WARNING: It generates a witness even if the internal var is constant, so it should be used only if the var is an input
+    /// to some ACIR opcode which requires a witness
+    pub(crate) fn get_or_compute_witness_unwrap(
+        &mut self,
+        mut var: InternalVar,
+        evaluator: &mut Evaluator,
+        ctx: &SsaContext,
+    ) -> Witness {
+        if let Some(v) = var.to_const() {
+            self.const_to_witness(v, evaluator, ctx)
+        } else {
+            let w = Self::internal_get_or_compute_witness(&mut var, evaluator)
+                .expect("infaillible non const expression");
+            var.set_witness(Some(w));
+            self.update(var);
+            w
+        }
+    }
+
+    /// Get or compute a witness equating the internal var
+    /// It returns None when the variable is a constant instead of creating a witness
+    /// because we should not need a witness in that case
+    /// If you really need one, you can use get_or_compute_witness_unwrap()
+    pub(crate) fn get_or_compute_witness(
+        &mut self,
+        mut var: InternalVar,
+        evaluator: &mut Evaluator,
+    ) -> Option<Witness> {
+        let w = Self::internal_get_or_compute_witness(&mut var, evaluator);
```

Comment:
Might be a bigger refactor than we want, just thinking out loud, if we can have the internalVarCache return an InternalVar that is immutable -- so the only mutable thing is the Cache